### DISCUSSION TITLE
Pygame

### DISCUSSION THREAD

**Discussion by seastar0687** [1101-GS4719-C/Python-Community-GS4719-C#3702] [General]
Pygame是一個可以用python來開發遊戲的package，可以繪製遊戲視窗，創建矩形，圓形等物件。
我個人使用後覺得如果要開發小遊戲pygame是一個非常適合的package
以下是範例

import pygame
import os
from Object import Object

# 初始化 Pygame
pygame.init()

# 設定遊戲視窗
W_H_ratio = 16 / 9
HEIGHT = 880
WIDTH = HEIGHT * W_H_ratio

class Song(Object):
    def __init__(self, position, title, picture=None, background=None):
        """
        初始化單首歌曲
        :param position: tuple, 歌曲的顯示位置 (x, y)
        :param title: str, 歌曲標題
        :param picture: pygame.Surface, 歌曲縮略圖
        :param background: pygame.Surface, 歌曲背景
        """
        super().__init__(position)
        self.title = title
        self.picture = picture
        self.background = background
        self.selected = False
        self.width = 800
        self.height = 120
        self.selected_width = 900
        self.selected_height = 160

        # 動畫屬性
        self.current_position = list(position)  # 當前位置
        self.target_position = list(position)  # 目標位置
        self.animation_speed = 10  # 每幀移動的像素量

    def update_position(self):
        """
        平滑更新當前位置，實現滑動動畫
        """
        for i in range(2):  # 更新 x 和 y 坐標
            if abs(self.current_position[i] - self.target_position[i]) < self.animation_speed:
                self.current_position[i] = self.target_position[i]
            else:
                if self.current_position[i] < self.target_position[i]:
                    self.current_position[i] += self.animation_speed
                elif self.current_position[i] > self.target_position[i]:
                    self.current_position[i] -= self.animation_speed

    def draw(self, surface, is_selected):
        """
        繪製歌曲物件
        :param surface: pygame.Surface, 畫布
        :param is_selected: bool, 是否被選中
        """
        transparent_surface = pygame.Surface((self.selected_width, self.selected_height), pygame.SRCALPHA)
        x, y = self.current_position
        if is_selected:
            pygame.draw.rect(transparent_surface, (50, 150, 250, 100), (0, 0, self.selected_width, self.selected_height))
            pygame.draw.rect(transparent_surface, (255, 255, 255), (0, 0, self.selected_width, self.selected_height), 3)
            surface.blit(transparent_surface, (x - 100, y))

            title_surface = pygame.font.Font("freesansbold.ttf", 32).render(self.title, True, (0, 0, 0))
            surface.blit(title_surface, (x - 100 + 30, y + self.selected_height / 2 - 10))
        else:
            pygame.draw.rect(transparent_surface, (200, 0, 200, 100), (0, 0, self.width, self.height))
            pygame.draw.rect(transparent_surface, (255, 255, 255), (0, 0, self.width, self.height), 3)
            surface.blit(transparent_surface, (x, y))

            title_surface = pygame.font.Font("freesansbold.ttf", 32).render(self.title, True, (255, 255, 255))
            surface.blit(title_surface, (x + 30, y + self.height / 2 - 10))

class SongSelector:
    def __init__(self, screen, songs_directory, font, width, height):
        """
        初始化歌曲選擇器
        """
        self.screen = screen
        self.songs_directory = songs_directory
        self.font = font
        self.width = width
        self.height = height
        self.songs = []
        self.key_pressed = {"up": False, "down": False}  # 上下鍵按壓狀態
        self.selected_index = 0
        self.running = True
        self._initialize_songs()

    def _initialize_songs(self):
        """
        加載歌曲資料並創建 Song 實例
        """
        song_list = os.listdir(self.songs_directory)
        y_position = self.height / 2 - 160 / 2  # 首個歌物件的起始 y 座標
        for i, song_name in enumerate(song_list):
            song_path = os.path.join(self.songs_directory, song_name)
            picture, background = 'No picture', 'No picture'

            # 檢查是否有圖片
            for content in os.listdir(song_path):
                if content.endswith('.jpg'):
                    image_path = os.path.join(song_path, content)
                    picture = pygame.image.load(image_path).convert_alpha()
                    pic_ratio = picture.get_width() / picture.get_height()
                    picture = pygame.transform.scale(picture, (300 * pic_ratio, 300))

                    background = pygame.image.load(image_path).convert_alpha()
                    background = pygame.transform.scale(background, (self.width, self.width / pic_ratio))
                    background.set_alpha(95)
                    break
            
            position = (self.width - 800 - 30, y_position + i * 140)
            self.songs.append(Song(position, song_name, picture, background))
        
        for i, song in enumerate(self.songs):
            if i == self.selected_index:
                song.target_position  = (song.position[0], self.height / 2 - song.selected_height / 2)
            elif i < self.selected_index:
                song.target_position  = (song.position[0], self.height / 2 - song.selected_height / 2 + (i - self.selected_index) * (song.height + 20) - 10)
            elif i > self.selected_index:
                song.target_position  = (song.position[0], self.height / 2 - song.selected_height / 2 + (i - self.selected_index - 1) * (song.height + 20) + song.selected_height + 20)

            

    def draw_background(self):
        """
        繪製當前選中歌曲的背景
        """
        if self.songs[self.selected_index].background != 'No picture':
            bg = self.songs[self.selected_index].background
            bg_ratio = bg.get_width() / bg.get_height()
            self.screen.blit(bg, (0, (self.height - self.width / bg_ratio) / 2))
        else:
            self.screen.fill((0, 0, 0))  # 默認黑色背景

    def handle_input(self, keys):
        """
        處理用戶輸入
        """

        if keys[pygame.K_x]:
            if not self.key_pressed["up"]:  # 檢測是否已處理
                if self.selected_index > 0:
                    self.selected_index -= 1
                    self._update_target_positions()
                self.key_pressed["up"] = True
        else:
            self.key_pressed["up"] = False  # 按鍵鬆開時重置狀態

        # 向下切換選項
        if keys[pygame.K_c]:
            if not self.key_pressed["down"]:  # 檢測是否已處理
                if self.selected_index < len(self.songs) - 1:
                    self.selected_index += 1
                    self._update_target_positions()
                self.key_pressed["down"] = True
        else:
            self.key_pressed["down"] = False  # 按鍵鬆開時重置狀態

    def _update_target_positions(self):
        # 更新每個 Song 的位置
        for i, song in enumerate(self.songs):
            if i == self.selected_index:
                song.target_position  = (song.position[0], self.height / 2 - song.selected_height / 2)
            elif i < self.selected_index:
                song.target_position  = (song.position[0], self.height / 2 - song.selected_height / 2 + (i - self.selected_index) * (song.height + 20) - 10)
            elif i > self.selected_index:
                song.target_position  = (song.position[0], self.height / 2 - song.selected_height / 2 + (i - self.selected_index - 1) * (song.height + 20) + song.selected_height + 20)

    def update_positions(self):
        """
        更新每首歌曲的當前位置
        """
        for song in self.songs:
            song.update_position()

    def draw_song_list(self):
        """
        繪製所有歌曲
        """
        for i, song in enumerate(self.songs):
            is_selected = (i == self.selected_index)
            song.draw(self.screen, is_selected)

    def update(self):
        """
        運行歌曲選擇邏輯
        """
        while self.running:
            keys = pygame.key.get_pressed()
            for event in pygame.event.get():
                if event.type == pygame.QUIT or keys[pygame.K_q]:
                    self.running = False

            self.handle_input(keys)
            self.update_positions()  # 更新滑動動畫
            self.draw_background()
            self.draw_song_list()
            pygame.display.update()

            if keys[pygame.K_z]:  # 確認選擇
                return self.selected_index

def main(): # 主函數
    screen = pygame.display.set_mode((WIDTH, HEIGHT))
    pygame.display.set_caption("Song Selector")
    pygame.display.set_caption("Song Selector with Transparent Rectangles")

    font = pygame.font.Font("game/image/Font.ttf", 32)
    songs_directory = "song"

    selector = SongSelector(screen, songs_directory, font, WIDTH, HEIGHT)
    selected_index = selector.update()
    return selected_index

if __name__ == "__main__":
    main()

**Comment by BOSENCHANG2004** [comment]
我覺得pygame這個套件如果掌握了拿來當作做GUI的的工具也是非常便利的!!

**Comment by BOSENCHANG2004** [comment]
我覺得 pygame 這個套件如果掌握了，拿來當作做 GUI 的工具也是非常便利的！
雖然 pygame 最初是為開發 2D 遊戲而設計，但它提供了完整的圖形繪製、事件處理、音效播放、字體渲染等功能。這些特性不僅適用於遊戲，也非常適合製作互動式的應用程式，像是小工具、教學軟體、模擬介面或資料可視化展示。
